### DISCUSSION TITLE
Terms like "prefix", "suffix", "boundaries", "entity" in completion.el

### DISCUSSION THREAD

**Discussion by dgutov** [company-mode/company-mode#1482] [General]
https://github.com/company-mode/company-mode/pull/1474 adopts more capabilities from CAPF, and adds new terms as well - in particular backend actions.

@monnier, @minad, (anybody else?), I wonder what you think about the added terminology.

In particular, I've added the backend action called `adjust-boundaries`, which roughly corresponds to `completion-boundaries` in minibuffer.el. It might be more succinct to call it just `boundaries`, but then the `prefix` action is extended to return a suffix string as well, which in turn calls for renaming the action - and the term "boundaries" seems to apply to it just as well.

`completion-at-point-functions` calls BEG and END "limits of entity" to complete - so I guess we have "entity limits", and we have "field boundaries" - one should note that the term "entity" doesn't appear in the docs outside of that docstring, and what is a "field" is not particularly defined either (except for file names, I guess). Would it be a good idea to use "entity" more, as a variable name and etc? I wonder.

And next we have what seems like the generalization of boundaries as `base-size` and `<name missing>` after the completions return from the completion style machinery. Could those also be referred to as "boundaries" of a sort? Or are they strictly "base size" and (maybe) "suffix size"?

Just thought to get some input before merging the above PR, which will codify these things somewhat rigidly in company-mode's APIs.

**Comment by monnier** [comment]
> In particular, I've added the backend action called `adjust-boundaries`,
> which roughly corresponds to `completion-boundaries` in minibuffer.el. It
> might be more succinct to call it just `boundaries`, but then the `prefix`
> action is extended to return a suffix string as well, which in turn calls
> for renaming the action - and the term "boundaries" seems to apply to it
> just as well.

I'm not super-happy with the term "boundaries".
And the BEG..END of `completion-at-point-functions` is something related
but different, as is the "base-size" of `completion-all-completions`.
So, yeah, it's a mess.

The BEG..END of `completion-at-point-functions` can be kept separate
from the other two, tho.  While "entity" is not defined and we could
improve its documentation, I think this BEG..END is usually well
understood by the relevant users, so I'm not too concerned about it.

The boundaries" are more problematic because users tend not to
understand them, largely because the vast majority of completion tables
don't need them, so most users don't even know there could be such
a need.

Over the years, I think I've learned that "boundaries" was a poor choice
because basically it describes the *type* of the return value but not
its meaning.  I think a better word would have been "subfield".

> And next we have what seems like the generalization of boundaries as
> `base-size` and `<name missing>` after the completions return from the
> completion style machinery. Could those also be referred to as "boundaries"
> of a sort? Or are they strictly "base size" and (maybe) "suffix size"?

You're talking here about the return value of
`completion-all-completions`, right?  In a sense these are just
a compression-mechanism, describing a part that is missing in each
listed completions because it was left unchanged.  "base-size" is
definitely a bad choice of word, so feel free to use another word.

Just as is the case for `completion-at-point-functions`, this part is
much less of a source of confusion, because the "base-size" is returned
together with the list of completions, so it's easier to describe what
it does.  A better API would probably return something like (BEG END
COMPLETIONS).

If we could merge `all-completions` and `completion-boundaries` (to
similarly return something of the form (BEG END COMPLETIONS)) it would
also make things more clear, but we need to know the boundaries in order
to compute the pattern to pass to `all-completions`, so it's hard to
merge them.

**Reply by dgutov** [reply]
> The BEG..END of `completion-at-point-functions` can be kept separate
from the other two, tho.  While "entity" is not defined and we could
improve its documentation, I think this BEG..END is usually well
understood by the relevant users, so I'm not too concerned about it.

Yeah, it's probably less of a problem for completion.el. Things are different for company-backends which has a separate call for this piece of data, so it needs a name (it's currently called "prefix").

> Over the years, I think I've learned that "boundaries" was a poor choice
because basically it describes the *type* of the return value but not
its meaning.

Right. Boundaries of _something_.

>  I think a better word would have been "subfield".

Subfield or just "field"? The term "completion field" is repeated throughout the package.

> You're talking here about the return value of
`completion-all-completions`, right?  In a sense these are just
a compression-mechanism, describing a part that is missing in each
listed completions because it was left unchanged.

Not just compression, although the effect is mostly visual. But it also determines the position of point after completion, for example. And the strings passed to `:exit-function`, annotation functions, etc.

> "base-size" is definitely a bad choice of word, so feel free to use another word.

Another word is exactly the thing I am looking for.

After going through the completion styles these are not "field boundaries" anymore, but something like the "entity boundaries that correspond to completion set", or "adjusted boundaries", as I've called them in the referenced PR.

**Comment by monnier** [comment]
>>  I think a better word would have been "subfield".
> Subfield or just "field"? The term "completion field" is repeated throughout the package.

For `completion-at-point-functions`, I'd say "field", and for
`completion-boundaries` I'd say "subfield".

>> You're talking here about the return value of
> `completion-all-completions`, right?  In a sense these are just
> a compression-mechanism, describing a part that is missing in each
> listed completions because it was left unchanged.
>
> Not just compression, although the effect is mostly visual. But it also
> determines the position of point after completion, for example. And the
> strings passed to `:exit-function`, annotation functions, etc.

Yes, the overall thing is not sufficiently well thought out, but
the idea is the same, which is to exclude the part that is not affected.

>> "base-size" is definitely a bad choice of word, so feel free to use another word.
> Another word is exactly the thing I am looking for.
> After going through the completion styles these are not "field boundaries"
> anymore, but something like the "entity boundaries that correspond to
> completion set", or "adjusted boundaries", as I've called them in the
> referenced PR.

"affected region"?

**Reply by dgutov** [reply]
> For `completion-at-point-functions`, I'd say "field", and for
`completion-boundaries` I'd say "subfield".

That makes a fair amount of sense, especially if we're coming from minibuffer completion.

> "affected region"?

Okay, hmm.

Like with "boundaries", "region" is a term for the type of the return value, so what remains is an adjective. For company-backends, so far it seems like a pair of strings is more immediately useful, so I'll probably stay with `adjust-boundaries` for now.

Thanks anyway, and maybe the above suggestions will work well in minibuffer.el.
